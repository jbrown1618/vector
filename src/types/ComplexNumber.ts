/**
 * Numbers of the form _a + bi_ where _i_ is the imaginary unit.
 */
export class ComplexNumber {
  public static readonly ZERO = new ComplexNumber(0, 0);
  public static readonly ONE = new ComplexNumber(1, 0);
  public static readonly I = new ComplexNumber(0, 1);

  private readonly _real: number;
  private readonly _imaginary: number;

  constructor(real: number, imaginary: number) {
    this._real = real;
    this._imaginary = imaginary;
  }

  public getRealPart(): number {
    return this._real;
  }

  public getImaginaryPart(): number {
    return this._imaginary;
  }

  public add(other: ComplexNumber): ComplexNumber {
    return new ComplexNumber(this._real + other._real, this._imaginary + other._imaginary);
  }

  public multiply(other: ComplexNumber): ComplexNumber {
    return new ComplexNumber(
      this._real * other._real - this._imaginary * other._imaginary,
      this._real * other._imaginary + this._imaginary * other._real
    );
  }

  public equals(other: ComplexNumber): boolean {
    return this._real === other._real && this._imaginary === other._imaginary;
  }

  public conjugate(): ComplexNumber {
    return new ComplexNumber(this._real, -1 * this._imaginary);
  }
}
