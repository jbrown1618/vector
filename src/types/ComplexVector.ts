import { ArrayVector } from './ArrayVector';
import { ComplexNumber } from './ComplexNumber';
import { VectorData } from './Vector';

export class ComplexVector extends ArrayVector<ComplexNumber> {
  constructor(data: VectorData<ComplexNumber>) {
    super(data);
  }

  protected newFromData(data: VectorData<ComplexNumber>): ArrayVector<ComplexNumber> {
    return new ComplexVector(data);
  }

  addScalars(first: ComplexNumber, second: ComplexNumber): ComplexNumber {
    return first.add(second);
  }

  conjugateScalar(scalar: ComplexNumber): ComplexNumber {
    return scalar.conjugate();
  }

  getAdditiveIdentity(): ComplexNumber {
    return ComplexNumber.ZERO;
  }

  getMultiplicativeIdentity(): ComplexNumber {
    return ComplexNumber.ONE;
  }

  multiplyScalars(first: ComplexNumber, second: ComplexNumber): ComplexNumber {
    return first.multiply(second);
  }

  scalarsEqual(first: ComplexNumber, second: ComplexNumber): boolean {
    return first.equals(second);
  }
}
