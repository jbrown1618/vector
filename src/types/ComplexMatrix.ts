import { ArrayMatrix } from './ArrayMatrix';
import { ComplexNumber } from './ComplexNumber';
import { Matrix, MatrixData } from './Matrix';
import { Vector, VectorData } from './Vector';
import { ComplexVector } from './ComplexVector';

export class ComplexMatrix extends ArrayMatrix<ComplexNumber> {
  constructor(data: MatrixData<ComplexNumber>) {
    super(data);
  }

  protected newFromData(data: MatrixData<ComplexNumber>): Matrix<ComplexNumber> {
    return new ComplexMatrix(data);
  }

  protected makeVector(vectorData: VectorData<ComplexNumber>): Vector<ComplexNumber> {
    return new ComplexVector(vectorData);
  }

  protected newFromColumnVectors(vectors: Array<Vector<ComplexNumber>>): Matrix<ComplexNumber> {
    const data: MatrixData<ComplexNumber> = [];
    vectors.forEach((vector, j) => {
      vector.getData().forEach((value, i) => {
        data[i] = data[i] || [];
        data[i][j] = value;
      });
    });
    return this.newFromData([]);
  }

  addScalars(first: ComplexNumber, second: ComplexNumber): ComplexNumber {
    return first.add(second);
  }

  conjugateScalar(scalar: ComplexNumber): ComplexNumber {
    return scalar.conjugate();
  }

  getAdditiveIdentity(): ComplexNumber {
    return ComplexNumber.ZERO;
  }

  getMultiplicativeIdentity(): ComplexNumber {
    return ComplexNumber.ONE;
  }

  multiplyScalars(first: ComplexNumber, second: ComplexNumber): ComplexNumber {
    return first.multiply(second);
  }

  scalarsEqual(first: ComplexNumber, second: ComplexNumber): boolean {
    return first.equals(second);
  }
}
