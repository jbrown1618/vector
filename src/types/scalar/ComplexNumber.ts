import { approximatelyEqual } from '../../utilities/NumberUtilities';

/**
 * Numbers of the form _a + bi_ where _i_ is the imaginary unit.
 */
export class ComplexNumber {
  public static readonly ZERO = new ComplexNumber(0, 0);
  public static readonly ONE = new ComplexNumber(1, 0);
  public static readonly NEG_ONE = new ComplexNumber(-1, 0);
  public static readonly I = new ComplexNumber(0, 1);

  private readonly _real: number;
  private readonly _imaginary: number;

  constructor(real: number, imaginary: number) {
    this._real = real;
    this._imaginary = imaginary;
  }

  public getRealPart(): number {
    return this._real;
  }

  public getImaginaryPart(): number {
    return this._imaginary;
  }

  public add(other: ComplexNumber): ComplexNumber {
    return new ComplexNumber(this._real + other._real, this._imaginary + other._imaginary);
  }

  public multiply(other: ComplexNumber): ComplexNumber {
    const ac = this._real * other._real;
    const bd = this._imaginary * other._imaginary;
    const aPlusBTimesCPlusD = (this._real + this._imaginary) * (other._real + other._imaginary);
    return new ComplexNumber(ac - bd, aPlusBTimesCPlusD - ac - bd);
  }

  public getAdditiveInverse(): ComplexNumber {
    return this.multiply(ComplexNumber.NEG_ONE);
  }

  public getMultiplicativeInverse(): ComplexNumber | undefined {
    const a = this.getRealPart();
    const b = this.getImaginaryPart();
    const denom = a * a + b * b;

    if (denom === 0) {
      return undefined;
    }

    const real = a / denom;
    const imag = (-1 * b) / denom;

    return new ComplexNumber(real, imag);
  }

  public equals(other: ComplexNumber): boolean {
    return (
      approximatelyEqual(this._real, other._real) &&
      approximatelyEqual(this._imaginary, other._imaginary)
    );
  }

  public conjugate(): ComplexNumber {
    return new ComplexNumber(this._real, -1 * this._imaginary);
  }
}
