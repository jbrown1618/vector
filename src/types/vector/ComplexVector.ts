import { ArrayVector } from './ArrayVector';
import { ComplexNumber } from '../scalar/ComplexNumber';
import { VectorConstructor, VectorData } from './Vector';
import { ComplexMatrix, VectorBuilder } from '../../index';
import { StaticImplements } from '../../utilities/StaticImplements';
import { ComplexNumberOperations } from '../scalar/ComplexNumberOperations';

/**
 * Implements `Vector` as an array of `ComplexNumber`s
 */
@StaticImplements<VectorConstructor<ComplexNumber, ComplexVector>>()
export class ComplexVector extends ArrayVector<ComplexNumber> {
  static ops(): ComplexNumberOperations {
    return new ComplexNumberOperations();
  }

  static builder(): VectorBuilder<ComplexNumber, ComplexVector> {
    return new VectorBuilder(ComplexVector);
  }

  ops(): ComplexNumberOperations {
    return ComplexVector.ops();
  }

  builder(): VectorBuilder<ComplexNumber, ComplexVector> {
    return ComplexVector.builder();
  }

  matrixBuilder() {
    return ComplexMatrix.builder();
  }

  constructor(data: VectorData<ComplexNumber>) {
    super(data);
  }
}
