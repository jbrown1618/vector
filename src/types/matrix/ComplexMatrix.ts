import { StaticImplements } from '../../utilities/StaticImplements';
import { ComplexNumber } from '../scalar/ComplexNumber';
import { ComplexNumberOperations } from '../scalar/ComplexNumberOperations';
import { ComplexVector } from '../vector/ComplexVector';
import { VectorBuilder } from '../vector/VectorBuilder';
import { ArrayMatrix } from './ArrayMatrix';
import { MatrixConstructor, MatrixData } from './Matrix';
import { MatrixBuilder } from './MatrixBuilder';
import { ScalarOperations } from '../scalar/ScalarOperations';

/**
 * Implements `Matrix` as a 2-dimensional array of `ComplexNumber`s
 * @public
 */
@StaticImplements<MatrixConstructor<ComplexNumber, ComplexVector, ComplexMatrix>>()
export class ComplexMatrix extends ArrayMatrix<ComplexNumber> {
  public static ops(): ScalarOperations<ComplexNumber> {
    return new ComplexNumberOperations();
  }

  public static builder(): MatrixBuilder<ComplexNumber, ComplexVector, ComplexMatrix> {
    return new MatrixBuilder(ComplexMatrix);
  }

  public static vectorBuilder(): VectorBuilder<ComplexNumber, ComplexVector> {
    return new VectorBuilder(ComplexVector);
  }

  constructor(data: MatrixData<ComplexNumber>) {
    super(data);
  }

  /**
   * {@inheritDoc TODO}
   */
  public ops(): ScalarOperations<ComplexNumber> {
    return ComplexMatrix.ops();
  }

  /**
   * {@inheritDoc TODO}
   */
  public builder(): MatrixBuilder<ComplexNumber, ComplexVector, ComplexMatrix> {
    return ComplexMatrix.builder();
  }

  /**
   * {@inheritDoc TODO}
   */
  public vectorBuilder(): VectorBuilder<ComplexNumber, ComplexVector> {
    return ComplexMatrix.vectorBuilder();
  }
}
