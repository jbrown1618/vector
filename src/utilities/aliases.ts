import { Vector } from '@lib/types/vector/Vector';
import { Matrix } from '@lib/types/matrix/Matrix';
import { NumberVector } from '@lib/types/vector/NumberVector';
import { NumberMatrix } from '@lib/types/matrix/NumberMatrix';

const vb = NumberVector.builder();
const mb = NumberMatrix.builder();

/**
 * Creates a new {@link Vector} of numbers.  See {@link VectorBuilder.fromArray}
 * @public
 */
export function vec(data: number[]): Vector<number> {
  return vb.fromArray(data);
}

/**
 * Creates a new {@link Matrix} of numbers.  See {@link MatrixBuilder.fromArray}
 * @public
 */
export function mat(data: number[][]): Matrix<number> {
  return mb.fromArray(data);
}

/**
 * Creates a new identity matrix of size `size`.  See {@link MatrixBuilder.identity}
 * @public
 */
export function eye(size: number): Matrix<number> {
  return mb.identity(size);
}

/**
 * Creates a new vector of all 1s.  See {@link VectorBuilder.ones}
 * @public
 */
export function ones(entries: number): Vector<number>;
/**
 * Creates a new matrix of all 1s.  See {@link MatrixBuilder.ones}
 * @public
 */
export function ones(rows: number, columns: number): Matrix<number>;
export function ones(rows: number, columns?: number): Matrix<number> | Vector<number> {
  return columns === undefined ? vb.ones(rows) : mb.ones(rows, columns);
}

/**
 * Creates a new vector of all 0s.  See {@link VectorBuilder.zeros}
 * @public
 */
export function zeros(entries: number): Vector<number>;
/**
 * Creates a new matrix of all 0s.  See {@link MatrixBuilder.zeros}
 * @public
 */
export function zeros(rows: number, columns: number): Matrix<number>;
export function zeros(rows: number, columns?: number): Matrix<number> | Vector<number> {
  return columns === undefined ? vb.zeros(rows) : mb.zeros(rows, columns);
}

/**
 * Creates a new matrix with the specified entries on the diagonal.  See {@link MatrixBuilder.diagonal}
 * @public
 */
export function diag(elements: number[]): Matrix<number> {
  return mb.diagonal(vec(elements));
}
