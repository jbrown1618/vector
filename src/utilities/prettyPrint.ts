import { Matrix } from '../types/matrix/Matrix';
import { Vector } from '../types/vector/Vector';

/**
 * @public
 */
export function prettyPrint(num: number): string;
export function prettyPrint<S>(vector: Vector<S>): string;
export function prettyPrint<S>(matrix: Matrix<S>): string;
export function prettyPrint<S>(input: number | Vector<S> | Matrix<S>): string {
  if (typeof input === 'number') {
    return prettyPrintNumber(input);
  }
  const isVector = !!(input as any).innerProduct;
  if (isVector) {
    return prettyPrintVector(input as Vector<S>);
  } else {
    return prettyPrintMatrix(input as Matrix<S>);
  }
}

function prettyPrintNumber(num: number): string {
  const str = num.toString();
  const fixed = num.toFixed(6);
  return str.length > fixed.length ? fixed : str;
}

function prettyPrintVector<S>(vector: Vector<S>): string {
  const ops = vector.ops();
  const width = getPrintWidth(vector);
  const prettyData = vector.toArray().map(val => ops.prettyPrint(val));
  return prettyData.map(value => '[ ' + spaces(width - value.length) + value + ' ]').join('\n');
}

function prettyPrintMatrix<S>(matrix: Matrix<S>): string {
  const ops = matrix.ops();
  const widthByCol = matrix.getColumnVectors().map(col => getPrintWidth(col));
  const prettyData = matrix.toArray().map(rowArr => rowArr.map(val => ops.prettyPrint(val)));

  return prettyData
    .map(
      rowArr =>
        '[ ' +
        rowArr.map((val, colIndex) => spaces(widthByCol[colIndex] - val.length) + val).join('  ') +
        ' ]'
    )
    .join('\n');
}

function getPrintWidth<S>(v: Vector<S>): number {
  return v
    .toArray()
    .map(val => v.ops().prettyPrint(val))
    .map(str => str.length)
    .reduce((max, curr) => (curr > max ? curr : max), 0);
}

function spaces(n: number): string {
  let sp = '';
  for (let i = 0; i < n; i++) {
    sp = sp + ' ';
  }
  return sp;
}
