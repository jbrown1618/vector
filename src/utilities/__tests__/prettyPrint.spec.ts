import { vec, mat } from '../aliases';
import { ComplexVector } from '../../types/vector/ComplexVector';
import { ComplexNumber } from '../../types/scalar/ComplexNumber';
import { ComplexMatrix } from '../../types/matrix/ComplexMatrix';
import { prettyPrint } from '../prettyPrint';

describe('prettyPrint', () => {
  test('prints a NumberVector', () => {
    const v = vec([1, 22, 333]);
    expect(prettyPrint(v)).toMatchSnapshot();

    const u = vec([1, 0.222, 1 / 3]);
    expect(prettyPrint(u)).toMatchSnapshot();
  });

  test('prints a ComplexVector', () => {
    const v = ComplexVector.builder().fromArray([
      new ComplexNumber(1, 1),
      new ComplexNumber(2, 22),
      new ComplexNumber(33, 333)
    ]);
    expect(prettyPrint(v)).toMatchSnapshot();
  });

  test('prints a NumberMatrix', () => {
    const v = mat([
      [1, 22222, 333],
      [444, 5, 6],
      [77, 88, 9999]
    ]);
    expect(prettyPrint(v)).toMatchSnapshot();
  });

  test('prints a ComplexMatrix', () => {
    const v = ComplexMatrix.builder().fromArray([
      [new ComplexNumber(1, 1), new ComplexNumber(22, 222), new ComplexNumber(33, 3)],
      [new ComplexNumber(44, 44), new ComplexNumber(5, 5), new ComplexNumber(6, 6)],
      [new ComplexNumber(7, 777), new ComplexNumber(888, 888), new ComplexNumber(9, 99999)]
    ]);
    expect(prettyPrint(v)).toMatchSnapshot();
  });
});
