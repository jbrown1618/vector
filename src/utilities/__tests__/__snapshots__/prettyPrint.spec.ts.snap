// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`prettyPrint prints a ComplexMatrix 1`] = `
"[   1 + 1i   22 + 222i     33 + 3i ]
[ 44 + 44i      5 + 5i      6 + 6i ]
[ 7 + 777i  888 + 888i  9 + 99999i ]"
`;

exports[`prettyPrint prints a ComplexVector 1`] = `
"[    1 + 1i ]
[   2 + 22i ]
[ 33 + 333i ]"
`;

exports[`prettyPrint prints a NumberMatrix 1`] = `
"[   1  22222   333 ]
[ 444      5     6 ]
[  77     88  9999 ]"
`;

exports[`prettyPrint prints a NumberVector 1`] = `
"[   1 ]
[  22 ]
[ 333 ]"
`;

exports[`prettyPrint prints a NumberVector 2`] = `
"[        1 ]
[    0.222 ]
[ 0.333333 ]"
`;
